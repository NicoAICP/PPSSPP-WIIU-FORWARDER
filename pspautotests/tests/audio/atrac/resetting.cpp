#include "shared.h"

void testBufferInfo(const char *title, int atracID, int sample, bool withInfo, u8 *basePtr) {
	AtracResetBufferInfo info;
	memset(&info, 0xCC, sizeof(info));

	checkpoint("%s: %08x", title, sceAtracGetBufferInfoForResetting(atracID, sample, withInfo ? &info : NULL));
	schedfResetBuffer(info, basePtr);
}

extern "C" int main(int argc, char *argv[]) {
	Atrac3File at3("sample.at3");
	at3.Require();
	LoadAtrac();

	int ignore;
	int atracID;
	u16 data[16384];
	
	atracID = sceAtracSetHalfwayBufferAndGetID((u8 *)at3.Data(), at3.Size() / 2, at3.Size() / 2);
	checkpointNext("IDs:");
	testBufferInfo("  Unallocated", 4, 0, true, at3.Data());
	testBufferInfo("  Invalid", -1, 0, true, at3.Data());
	testBufferInfo("  Valid", atracID, 0, true, at3.Data());
	
	checkpointNext("Sample values:");
	const static int samples[] = {0, 1, 1679, 1680, 2046, 2047, 0x1000, 0x47ff, 0x10000, 247499, 247500, 247501, 0x100000, 0x1000000, 0x7FFFFFFF, -1, -2, -2048, -4096};
	for (size_t i = 0; i < ARRAY_SIZE(samples); ++i) {
		char temp[128];
		snprintf(temp, 128, "  %08x", samples[i]);
		testBufferInfo(temp, atracID, samples[i], true, at3.Data());
	}

	checkpointNext("Other states:");
	int withoutDataID = sceAtracGetAtracID(0x1000);
	testBufferInfo("  No data", withoutDataID, 0, true, at3.Data());
	sceAtracReleaseAtracID(withoutDataID);
	forceAtracState(atracID, 8);
	testBufferInfo("  State 8", atracID, 0, true, at3.Data());
	forceAtracState(atracID, 16);
	testBufferInfo("  State 16", atracID, 0, true, at3.Data());
	forceAtracState(atracID, 1);
	sceAtracReleaseAtracID(atracID);

	// Crashes.
	//checkpointNext("Buffer:");
	//testBufferInfo("  NULL", atracID, 0, false, at3.Data());

	at3.Reload("sample.at3");
	atracID = sceAtracSetDataAndGetID(at3.Data(), at3.Size());
	checkpointNext("Entire buffer:");
	testBufferInfo("  At start -> 0", atracID, 0, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 1 -> 0", atracID, 0, true, at3.Data());

	testBufferInfo("  After decode 1 -> 65536", atracID, 65536, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 2 -> 65536", atracID, 65536, true, at3.Data());

	testBufferInfo("  After decode 2 -> 196608", atracID, 196608, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 3 -> 196608", atracID, 196608, true, at3.Data());

	sceAtracReleaseAtracID(atracID);
	at3.Reload("sample.at3");
	atracID = sceAtracSetHalfwayBufferAndGetID((u8 *)at3.Data(), at3.Size() / 2, at3.Size());
	checkpointNext("Half buffer:");
	testBufferInfo("  At start -> 0", atracID, 0, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 1 -> 0", atracID, 0, true, at3.Data());

	testBufferInfo("  After decode 1 -> 65536", atracID, 65536, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 2 -> 65536", atracID, 65536, true, at3.Data());

	testBufferInfo("  After decode 2 -> 196608", atracID, 196608, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 3 -> 196608", atracID, 196608, true, at3.Data());

	sceAtracReleaseAtracID(atracID);
	at3.Reload("sample.at3");
	atracID = sceAtracSetHalfwayBufferAndGetID((u8 *)at3.Data(), at3.Size() / 2, at3.Size() / 2);
	checkpointNext("Streamed buffer:");
	testBufferInfo("  At start -> 0", atracID, 0, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 1 -> 0", atracID, 0, true, at3.Data());

	testBufferInfo("  After decode 1 -> 65536", atracID, 65536, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 2 -> 65536", atracID, 65536, true, at3.Data());

	testBufferInfo("  After decode 2 -> 196608", atracID, 196608, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 3 -> 196608", atracID, 196608, true, at3.Data());

	return 0;
}
