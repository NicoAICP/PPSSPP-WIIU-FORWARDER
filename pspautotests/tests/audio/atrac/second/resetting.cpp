#include "../shared.h"

void testBufferInfo(const char *title, int atracID, int sample, u8 *basePtr) {
	AtracResetBufferInfo info;
	memset(&info, 0xCC, sizeof(info));

	checkpoint("%s: %08x", title, sceAtracGetBufferInfoForResetting(atracID, sample, &info));
	schedfResetBuffer(info, basePtr);
}

void testDirectResetPlayPos(const char *title, int atracID, int sample, int firstBytes, int secondBytes) {
	checkpoint("%s: %08x", title, sceAtracResetPlayPosition(atracID, sample, firstBytes, secondBytes));
}

void testSimpleResetPlayPos(const char *title, int atracID, int sample, u8 *input, Atrac3File &at3) {
	AtracResetBufferInfo info = {};
	if (sceAtracGetBufferInfoForResetting(atracID, sample, &info) == 0) {
		if (info.first.minWriteBytes != 0) {
			memcpy(info.first.writePos, at3.Data() + info.first.filePos, info.first.minWriteBytes);
		}
		testDirectResetPlayPos(title, atracID, sample, info.first.minWriteBytes, 0);
	} else {
		testDirectResetPlayPos(title, atracID, sample, 0, 0);
	}
}

static void testResetInfoFuncs(int atracID, Atrac3File &at3, u8 *basePtr) {
	testBufferInfo("  Reset to start", atracID, 0, basePtr);
	testBufferInfo("  Reset before loop", atracID, 200000 - 2048 * 5, basePtr);
	testBufferInfo("  Reset to loop", atracID, 200000, basePtr);
	testBufferInfo("  Reset end of loop frame", atracID, 200335 + 368, basePtr);
	testBufferInfo("  Reset after loop", atracID, 200335 + 369, basePtr);
	testBufferInfo("  Reset to end", atracID, 247500, basePtr);
}

static void testResetPosFuncs(int atracID, Atrac3File &at3, u8 *basePtr) {
	testSimpleResetPlayPos("  Play at start", atracID, 0, basePtr, at3);
	testSimpleResetPlayPos("  Play before loop", atracID, 200000 - 2048 * 5, basePtr, at3);
	testSimpleResetPlayPos("  Play at loop", atracID, 200000, basePtr, at3);
	testSimpleResetPlayPos("  Play at end of loop frame", atracID, 200335 + 368, basePtr, at3);
	testSimpleResetPlayPos("  Play after loop", atracID, 200335 + 369, basePtr, at3);
	testSimpleResetPlayPos("  Play at end", atracID, 247500, basePtr, at3);
}

extern "C" int main(int argc, char *argv[]) {
	LoadAtrac();

	Atrac3File at3("../sample.at3");
	at3.Require();

	Atrac3File looped((size_t)0);
	CreateLoopedAtracFrom(at3, looped, 2048, 200000);

	u8 *buffer = new u8[looped.Size() / 2];
	memcpy(buffer, looped.Data(), looped.Size() / 2);

	int atracID = sceAtracSetDataAndGetID(buffer, looped.Size() / 2);

	if (sceAtracIsSecondBufferNeeded(atracID) != 1) {
		printf("FATAL: Does not need second buffer.\n");
		sceAtracReleaseAtracID(atracID);
		delete [] buffer;
		return 1;
	}

	checkpointNext("Reset info without second buffer:");
	testResetInfoFuncs(atracID, looped, buffer);
	checkpointNext("Reset pos without second buffer:");
	testResetPosFuncs(atracID, looped, buffer);

	u32 off = 0, size = 0;
	sceAtracGetSecondBufferInfo(atracID, &off, &size);
	u8 *secondBuffer = new u8[size];
	memcpy(secondBuffer, looped.Data() + off, size);

	sceAtracSetSecondBuffer(atracID, secondBuffer, 368 * 5);

	checkpointNext("Reset info with partial second buffer:");
	testResetInfoFuncs(atracID, looped, buffer);
	checkpointNext("Reset pos with partial second buffer:");
	testResetPosFuncs(atracID, looped, buffer);

	sceAtracSetSecondBuffer(atracID, secondBuffer, size);

	checkpointNext("Reset info with full second buffer:");
	testResetInfoFuncs(atracID, looped, buffer);
	checkpointNext("Reset pos with full second buffer:");
	testResetPosFuncs(atracID, looped, buffer);

	sceAtracReleaseAtracID(atracID);

	delete [] buffer;
	delete [] secondBuffer;

	UnloadAtrac();
	return 0;
}