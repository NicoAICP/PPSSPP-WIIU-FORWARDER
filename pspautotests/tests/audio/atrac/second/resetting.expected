[x] Reset info without second buffer:
[x]   Reset to start: 80630012
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   Reset before loop: 80630012
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   Reset to loop: 80630012
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   Reset end of loop frame: 80630012
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   Reset after loop: 80630012
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   Reset to end: 80630012
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc

[x] Reset pos without second buffer:
[x]   Play at start: 80630012
[x]   Play before loop: 80630012
[x]   Play at loop: 80630012
[x]   Play at end of loop frame: 80630012
[x]   Play after loop: 80630012
[x]   Play at end: 80630012

[x] Reset info with partial second buffer:
[x]   Reset to start: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=000000a4
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset before loop: 00000000
   #1: write=p+0x0, writable=000008d0, min=000002f0, file=000087c4
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset to loop: 00000000
   #1: write=p+0x0, writable=00000178, min=00000178, file=00008f1c
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset end of loop frame: 00000000
   #1: write=p+0x0, writable=00000178, min=00000178, file=00008f1c
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset after loop: 00000000
   #1: write=p+0x0, writable=00001ed8, min=00000000, file=00009674
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset to end: 00000000
   #1: write=p+0x0, writable=00000468, min=00000468, file=0000b0e4
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000

[x] Reset pos with partial second buffer:
[r]   Play at start: 00000000
[r]   Play before loop: 00000000
[r]   Play at loop: 00000000
[r]   Play at end of loop frame: 00000000
[r]   Play after loop: 00000000
[r]   Play at end: 00000000

[x] Reset info with full second buffer:
[x]   Reset to start: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=000000a4
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset before loop: 00000000
   #1: write=p+0x0, writable=000008d0, min=000002f0, file=000087c4
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset to loop: 00000000
   #1: write=p+0x0, writable=00000178, min=00000178, file=00008f1c
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset end of loop frame: 00000000
   #1: write=p+0x0, writable=00000178, min=00000178, file=00008f1c
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset after loop: 00000000
   #1: write=p+0x0, writable=00000000, min=00000000, file=00000000
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   Reset to end: 00000000
   #1: write=p+0x0, writable=00000000, min=00000000, file=00000000
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000

[x] Reset pos with full second buffer:
[r]   Play at start: 00000000
[r]   Play before loop: 00000000
[r]   Play at loop: 00000000
[r]   Play at end of loop frame: 00000000
[r]   Play after loop: 00000000
[r]   Play at end: 00000000
